File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        ans = []
        
        def bfs(start):
            queue, visited = deque([start]), set()
            visited.add(start)
            
            while queue:
                total = 0
                n = len(queue)
                for i in range(n):
                    cur = queue.popleft()
                    total += cur.val
                    if cur.left:
                        queue.append(cur.left)
                    if cur.right:
                        queue.append(cur.right)
                ans.append(total/n)                                      
                total = 0
        bfs(root)
        
        return ans

File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        queue<TreeNode*> q;
        vector<double> res;
        q.push(root);
        
        while (!q.empty()) {
            int n = q.size();
            double cur = 0;
            
            for (int i = 0; i  < n; i++) {
                auto u = q.front();
                cur += u->val;
                
                if (u->left)
                    q.push(u->left);
                
                if (u->right)
                    q.push(u->right);
                q.pop();
            }
            res.push_back(cur / n);
        }
        return res;
    }
};